File: Week_3/Day_5/findDIamaterBT.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.maxD = 0
        self.findMaxDiam(root)
        return self.maxD
    
    def findMaxDiam(self, root):
        if not root: return 0
        left = self.findMaxDiam(root.left)
        right = self.findMaxDiam(root.right)
        diameter = left + right
        self.maxD = max(self.maxD, diameter)
        return max(left, right) + 1

File: Week_3/Day_5/findDiamaterBT.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function (root) {
    if (!root) {
        return 0
    }
    let maxD = 0
    const findMaxDiam = (node) => {
        if (!node) return 0
        let left = findMaxDiam(node.left)
        let right = findMaxDiam(node.right)
        maxD = Math.max(maxD, left + right)
        return Math.max(left, right) + 1
    }
    findMaxDiam(root)
    return maxD

};